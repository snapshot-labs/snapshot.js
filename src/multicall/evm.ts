import {
  Fragment,
  FunctionFragment,
  Interface,
  JsonFragment
} from '@ethersproject/abi';
import { Signer } from '@ethersproject/abstract-signer';
import { Contract } from '@ethersproject/contracts';
import { Provider } from '@ethersproject/providers';

const multicallAbi = [
  'function aggregate(tuple(address target, bytes callData)[] calls) view returns (uint256 blockNumber, bytes[] returnData)'
];

export default async function multicall(
  address: string,
  provider: Signer | Provider,
  abi: string | (Fragment | JsonFragment | string)[],
  calls: [string, FunctionFragment | string, any[]][],
  limit: number,
  options = {}
): Promise<any> {
  const multi = new Contract(address, multicallAbi, provider);
  const itf = new Interface(abi);
  try {
    const pages = Math.ceil(calls.length / limit);
    const promises: any = [];
    Array.from(Array(pages)).forEach((x, i) => {
      const callsInPage = calls.slice(limit * i, limit * (i + 1));
      promises.push(
        multi.aggregate(
          callsInPage.map((call) => [
            call[0].toLowerCase(),
            itf.encodeFunctionData(call[1], call[2])
          ]),
          options
        )
      );
    });
    let results: any[] = await Promise.all(promises);
    results = results.reduce((prev: any, [, res]: any) => prev.concat(res), []);
    return results.map((call, i) =>
      itf.decodeFunctionResult(calls[i][1], call)
    );
  } catch (e: any) {
    return Promise.reject(e);
  }
}
