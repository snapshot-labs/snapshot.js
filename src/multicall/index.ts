import multicallEvm from './evm';
import multicallStarknet from './starknet';
import Multicaller from './multicaller';
import networks from '../networks.json';

const MULTICALLS_FN = {
  evm: multicallEvm,
  starknet: multicallStarknet
} as const;

async function multicall(
  network: string,
  provider,
  abi: any[],
  calls: any[],
  options: Record<string, any> = {}
) {
  const address = options?.multicallAddress || networks[network].multicall;

  if (!address) {
    throw new Error('missing multicall address');
  }

  const multicallOptions = { ...options };
  const limit = multicallOptions?.limit || 500;

  delete multicallOptions.limit;
  delete multicallOptions.multicallAddress;

  const protocol = networks[network]?.starknet ? 'starknet' : 'evm';

  return MULTICALLS_FN[protocol](
    address,
    provider,
    abi,
    calls,
    limit,
    multicallOptions
  );
}

export { multicall, Multicaller };
