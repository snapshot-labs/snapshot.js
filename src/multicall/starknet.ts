import { num, RpcProvider, shortString, transaction, uint256 } from 'starknet';

/**
 * Parses the raw result from a Starknet function call based on its ABI.
 * It handles different types like felt252, u8, u256, etc., and decodes them accordingly.
 * @param rawResult - The raw result from the Starknet function call.
 * @param functionAbi - The ABI of the function that was called.
 * @returns The parsed result in a more usable format.
 */
function parseStarknetResult(rawResult: string[], functionAbi: any): any {
  if (
    !functionAbi ||
    !functionAbi.outputs ||
    !Array.isArray(rawResult) ||
    rawResult.length === 0
  ) {
    return rawResult;
  }

  // Parse each output according to its type
  const outputs = functionAbi.outputs;
  const results: any[] = [];
  let rawIndex = 0; // Track position in rawResult array

  try {
    for (let outputIndex = 0; outputIndex < outputs.length; outputIndex++) {
      const output = outputs[outputIndex];
      const rawValue = rawResult[rawIndex];

      switch (output.type) {
        case 'core::felt252':
          try {
            results.push(shortString.decodeShortString(rawValue));
          } catch {
            results.push(rawValue);
          }
          rawIndex++;
          break;
        case 'core::integer::u8':
        case 'core::integer::u16':
        case 'core::integer::u32':
        case 'core::integer::u64':
          results.push(parseInt(rawValue, 16));
          rawIndex++;
          break;
        case 'core::integer::u128':
        case 'core::integer::usize':
          results.push(BigInt(rawValue).toString());
          rawIndex++;
          break;
        case 'core::integer::u256':
          results.push(
            uint256.uint256ToBN({
              low: rawValue,
              high: rawResult[rawIndex + 1] || '0x0'
            })
          );
          rawIndex += 2; // u256 uses two slots
          break;
        case 'core::integer::i8':
        case 'core::integer::i16':
        case 'core::integer::i32':
        case 'core::integer::i64':
          results.push(parseInt(rawValue, 16));
          rawIndex++;
          break;
        case 'core::integer::i128':
          results.push(BigInt(rawValue).toString());
          rawIndex++;
          break;
        case 'core::bool':
          results.push(rawValue === '0x1' || rawValue === '0x01');
          rawIndex++;
          break;
        case 'core::starknet::contract_address::ContractAddress':
        case 'core::starknet::class_hash::ClassHash':
        case 'core::starknet::storage_access::StorageAddress':
          results.push(rawValue);
          rawIndex++;
          break;
        case 'core::bytes_31::bytes31':
          results.push(rawValue);
          rawIndex++;
          break;
        default:
          results.push(rawValue);
          rawIndex++;
      }
    }
    return results;
  } catch {
    return rawResult;
  }
}

/**
 * Partitions the responses from a Starknet multicall into individual call results.
 * Each response starts with its length, followed by the actual response data.
 * @param responses - The array of responses from the Starknet multicall.
 * @returns An array of arrays, where each inner array contains the response data for a single call.
 */
const partitionResponses = (responses: string[]): string[][] => {
  if (responses.length === 0) {
    return [];
  }

  const [responseLength, ...restResponses] = responses;
  const responseLengthInt = Number(num.toBigInt(responseLength));
  const response = restResponses.slice(0, responseLengthInt);
  const remainingResponses = restResponses.slice(responseLengthInt);

  return [response, ...partitionResponses(remainingResponses)];
};

export default async function multicall(
  address: string,
  provider: RpcProvider,
  abi: any[],
  calls: any[],
  limit: number,
  options: Record<string, any> = {}
) {
  const callData = calls.map((call) => {
    return {
      contractAddress: call[0],
      entrypoint: call[1],
      calldata: call[2] || []
    };
  });

  // Chunk calls into batches based on limit
  const chunks: any[][] = [];
  for (let i = 0; i < callData.length; i += limit) {
    chunks.push(callData.slice(i, i + limit));
  }

  // Process each chunk
  const paginatedResults = await Promise.all(
    chunks.map((chunk) =>
      provider.callContract(
        {
          contractAddress: address,
          entrypoint: 'aggregate',
          calldata: transaction.fromCallsToExecuteCalldata(chunk)
        },
        options.blockTag ?? 'latest'
      )
    )
  );

  const callResults = paginatedResults
    .map((callContractResult) => {
      // eslint-disable-next-line @typescript-eslint/no-unused-vars
      const [_blockNumber, _totalLength, ...results] = callContractResult;
      return partitionResponses(results);
    })
    .flat();

  return callResults.map((result, index) => {
    const [, functionName] = calls[index];
    const functionAbi = abi.find((item) => item.name === functionName);

    const parsedResult = parseStarknetResult(result, functionAbi);
    return parsedResult;
  });
}
