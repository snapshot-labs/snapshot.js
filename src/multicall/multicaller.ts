import { StaticJsonRpcProvider } from '@ethersproject/providers';
import set from 'lodash.set';
import { RpcProvider } from 'starknet';
import { multicall } from './';

type Path = string | number | number[] | string[];

export default class Multicaller {
  public network: string;
  public provider: StaticJsonRpcProvider | RpcProvider;
  public abi: any[];
  public options: any = {};
  public calls: any[] = [];
  public paths: Path[] = [];

  constructor(
    network: string,
    provider: StaticJsonRpcProvider | RpcProvider,
    abi: any[],
    options: any = {}
  ) {
    this.network = network;
    this.provider = provider;
    this.abi = abi;
    this.options = options;
  }

  call(path: Path, address: string, fn: string, params?: any[]): Multicaller {
    this.calls.push([address, fn, params]);
    this.paths.push(path);
    return this;
  }

  async execute(from?: any): Promise<any> {
    const obj = from || {};
    const result = await multicall(
      this.network,
      this.provider,
      this.abi,
      this.calls,
      this.options
    );
    result.forEach((r: any, i: number) =>
      set(obj, this.paths[i], r.length > 1 ? r : r[0])
    );
    this.calls = [];
    this.paths = [];
    return obj;
  }
}
