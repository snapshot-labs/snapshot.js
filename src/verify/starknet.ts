import { Contract, RpcProvider, typedData } from 'starknet';
import type { SignaturePayload } from '.';
import type { ProviderOptions } from '../utils/provider';

export type NetworkType = 'SN_MAIN' | 'SN_SEPOLIA';

const RPC_URLS: Record<NetworkType, string> = {
  SN_MAIN: 'https://starknet-mainnet.public.blastapi.io',
  SN_SEPOLIA: 'https://starknet-sepolia.public.blastapi.io'
};

const ABI = [
  {
    name: 'argent::account::interface::IDeprecatedArgentAccount',
    type: 'interface',
    items: [
      {
        name: 'isValidSignature',
        type: 'function',
        inputs: [
          {
            name: 'hash',
            type: 'core::felt252'
          },
          {
            name: 'signatures',
            type: 'core::array::Array::<core::felt252>'
          }
        ],
        outputs: [
          {
            type: 'core::felt252'
          }
        ],
        state_mutability: 'view'
      }
    ]
  }
];

function getProvider(network: NetworkType, options: ProviderOptions) {
  if (!RPC_URLS[network]) throw new Error('Invalid network');

  return new RpcProvider({
    nodeUrl: options?.broviderUrl ?? RPC_URLS[network]
  });
}

export function isStarknetMessage(data: SignaturePayload): boolean {
  return !!data.primaryType && !!data.types.StarkNetDomain;
}

export function getHash(data: SignaturePayload, address: string): string {
  const { domain, types, primaryType, message } =
    data as Required<SignaturePayload>;

  return typedData.getMessageHash(
    { types, primaryType, domain, message },
    address
  );
}

export default async function verify(
  address: string,
  sig: string[],
  data: SignaturePayload,
  network: NetworkType = 'SN_MAIN',
  options: ProviderOptions = {}
): Promise<boolean> {
  const contractAccount = new Contract(
    ABI,
    address,
    getProvider(network, options)
  );

  await contractAccount.isValidSignature(getHash(data, address), [
    sig[0],
    sig[1]
  ]);

  return true;
}
