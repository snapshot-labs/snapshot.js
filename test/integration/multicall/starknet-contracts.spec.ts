// To run this test:
// npm test -- --run test/integration/multicall/starknet-contracts.spec.ts

import { describe, it, expect } from 'vitest';
import { RpcProvider } from 'starknet';
import multicall from '../../../src/multicall/starknet';

describe('Starknet Contract Reader Integration', () => {
  // Contract addresses
  const STAKING_DELEGATION_POOL =
    '0x02cb02c72e8a0975e69e88298443e984d965a49eab38f5bdde1f5072daa09cfe';

  const STARKNET_STAKING =
    '0x00ca1702e64c81d9a07b86bd2c540188d92a2c73cf5cc0e508d949015e7e84a7';

  // Multicall contract address for Starknet mainnet
  const MULTICALL_ADDRESS =
    '0x05754af3760f3356da99aea5c3ec39ccac7783d925a19666ebbeca58ff0087f4';

  // Input parameters
  const POOL_MEMBER_ADDRESS =
    '0x01c6a9f1f3f97a0a3b176fac97b69c3cd21ba31a522fca111b0e49e46975a048';

  const STAKER_ADDRESS =
    '0x00d3b910d8c528bf0216866053c3821ac6c97983dc096bff642e9a3549210ee7';

  // Initialize provider
  const provider = new RpcProvider({
    nodeUrl: 'https://starknet-mainnet.public.blastapi.io/rpc/v0_7'
  });

  /**
   * Format contract response data with proper JSON formatting
   */
  function formatOutput(title: string, data: any): void {
    console.log(`\n${'='.repeat(60)}`);
    console.log(`${title}`);
    console.log(`${'='.repeat(60)}`);

    if (typeof data === 'object' && data !== null) {
      console.log(JSON.stringify(data, null, 2));
    } else {
      console.log(data);
    }

    console.log(`${'='.repeat(60)}\n`);
  }

  /**
   * Convert hex response to readable format
   */
  function convertHexToString(hexValue: string): string {
    if (typeof hexValue === 'string' && hexValue.startsWith('0x')) {
      try {
        const hex = hexValue.slice(2);
        let str = '';
        for (let i = 0; i < hex.length; i += 2) {
          const charCode = parseInt(hex.substr(i, 2), 16);
          if (charCode > 0) {
            str += String.fromCharCode(charCode);
          }
        }
        return str || hexValue;
      } catch (e) {
        return hexValue;
      }
    }
    return hexValue;
  }

  /**
   * Contract ABI definitions for the functions we're calling
   */
  const contractAbi = [
    {
      name: 'identify',
      outputs: [{ type: 'core::felt252' }]
    },
    {
      name: 'get_pool_member_info_v1',
      outputs: [
        { type: 'core::starknet::contract_address::ContractAddress' }, // reward_address
        { type: 'core::integer::u256' }, // amount
        { type: 'core::integer::u256' }, // unclaimed_rewards
        { type: 'core::integer::u256' }, // commission
        { type: 'core::integer::u256' }, // unpool_amount
        { type: 'core::integer::u64' }, // unpool_time
        { type: 'core::integer::u8' } // status
      ]
    },
    {
      name: 'get_staker_info_v1',
      outputs: [
        { type: 'core::starknet::contract_address::ContractAddress' }, // reward_address
        { type: 'core::starknet::contract_address::ContractAddress' }, // operational_address
        { type: 'core::integer::u64' }, // unstake_time
        { type: 'core::integer::u8' }, // amount_own
        { type: 'core::integer::u256' }, // amount_delegated
        { type: 'core::integer::u256' }, // unclaimed_rewards_own
        { type: 'core::integer::u64' }, // commission
        { type: 'core::starknet::contract_address::ContractAddress' }, // pool_contract
        { type: 'core::integer::u256' }, // unclaimed_rewards_delegated
        { type: 'core::integer::u64' } // switch_pool_time
      ]
    }
  ];

  it('should call identify, get_pool_member_info_v1, and get_staker_info_v1 using multicall', async () => {
    console.log('\nüöÄ Starting Starknet Contract Reader with Multicall...\n');
    console.log('üìä Contract Information:');
    console.log(`   ‚Ä¢ Multicall Contract:      ${MULTICALL_ADDRESS}`);
    console.log(`   ‚Ä¢ Staking Delegation Pool: ${STAKING_DELEGATION_POOL}`);
    console.log(`   ‚Ä¢ Starknet Staking:        ${STARKNET_STAKING}`);
    console.log(`   ‚Ä¢ Pool Member Address:     ${POOL_MEMBER_ADDRESS}`);
    console.log(`   ‚Ä¢ Staker Address:          ${STAKER_ADDRESS}\n`);

    // Prepare multicall calls
    const calls = [
      // Call 1: identify() on Staking Delegation Pool
      [STAKING_DELEGATION_POOL, 'identify', []],

      // Call 2: get_pool_member_info_v1() on Staking Delegation Pool
      [
        STAKING_DELEGATION_POOL,
        'get_pool_member_info_v1',
        [POOL_MEMBER_ADDRESS]
      ],

      // Call 3: get_staker_info_v1() on Starknet Staking
      [STARKNET_STAKING, 'get_staker_info_v1', [STAKER_ADDRESS]]
    ];

    // Execute multicall
    console.log('üîÑ Executing multicall...\n');
    const results = await multicall(
      MULTICALL_ADDRESS,
      provider,
      contractAbi,
      calls,
      10 // batch limit
    );

    // Validate that we got results
    expect(results).toBeDefined();
    expect(results).toHaveLength(3);

    // Process results
    const [identifyResult, poolMemberInfoResult, stakerInfoResult] = results;

    // Validate identify result
    expect(identifyResult).toBeDefined();
    expect(identifyResult.length).toBeGreaterThan(0);

    // Format identify result
    const identityString = convertHexToString(identifyResult[0]);
    formatOutput('üîç STAKING DELEGATION POOL - IDENTIFY', {
      contract: STAKING_DELEGATION_POOL,
      function: 'identify()',
      raw_response: identifyResult[0],
      decoded_response: identityString,
      timestamp: new Date().toISOString()
    });

    // Validate pool member info result
    expect(poolMemberInfoResult).toBeDefined();
    expect(poolMemberInfoResult.length).toBeGreaterThanOrEqual(6);

    // Format pool member info result
    formatOutput('üë• STAKING DELEGATION POOL - MEMBER INFO', {
      contract: STAKING_DELEGATION_POOL,
      function: 'get_pool_member_info_v1(member_address)',
      input: {
        member_address: POOL_MEMBER_ADDRESS
      },
      response: {
        reward_address: poolMemberInfoResult[0] || '0x0',
        amount: poolMemberInfoResult[1]?.toString() || '0x0',
        unclaimed_rewards: poolMemberInfoResult[2]?.toString() || '0x0',
        commission: poolMemberInfoResult[3]?.toString() || '0x0',
        unpool_amount: poolMemberInfoResult[4]?.toString() || '0x0',
        unpool_time: poolMemberInfoResult[5] || 0,
        status: poolMemberInfoResult[6] || 0
      },
      timestamp: new Date().toISOString()
    });

    // Validate staker info result
    expect(stakerInfoResult).toBeDefined();
    expect(stakerInfoResult.length).toBeGreaterThanOrEqual(5);

    // Format staker info result
    formatOutput('üèõÔ∏è  STARKNET STAKING - STAKER INFO', {
      contract: STARKNET_STAKING,
      function: 'get_staker_info_v1(staker_address)',
      input: {
        staker_address: STAKER_ADDRESS
      },
      response: {
        staker_address: STAKER_ADDRESS,
        reward_address: stakerInfoResult[0] || '0x0',
        operational_address: stakerInfoResult[1] || '0x0',
        unstake_time: stakerInfoResult[2] || 0,
        amount_own: stakerInfoResult[3] || 0,
        amount_delegated: stakerInfoResult[4]?.toString() || '0x0',
        unclaimed_rewards_own: stakerInfoResult[5]?.toString() || '0x0',
        commission: stakerInfoResult[6] || 0,
        pool_contract: stakerInfoResult[7] || '0x0',
        unclaimed_rewards_delegated: stakerInfoResult[8]?.toString() || '0x0',
        switch_pool_time: stakerInfoResult[9] || 0
      },
      timestamp: new Date().toISOString()
    });

    // Summary
    console.log('\nüìã EXECUTION SUMMARY');
    console.log('='.repeat(60));
    console.log(`‚úÖ identify(): ${identifyResult ? 'SUCCESS' : 'FAILED'}`);
    console.log(
      `‚úÖ get_pool_member_info_v1(): ${
        poolMemberInfoResult ? 'SUCCESS' : 'FAILED'
      }`
    );
    console.log(
      `‚úÖ get_staker_info_v1(): ${stakerInfoResult ? 'SUCCESS' : 'FAILED'}`
    );
    console.log('='.repeat(60));
    console.log(
      '\nüéâ All contract calls completed successfully via multicall!'
    );

    // Final assertions
    expect(identityString).toBe('Staking Delegation Pool');
    expect(poolMemberInfoResult[0]).toBeDefined(); // reward_address exists
    expect(stakerInfoResult[0]).toBeDefined(); // reward_address exists
  }, 30000); // 30 second timeout for network calls
});
