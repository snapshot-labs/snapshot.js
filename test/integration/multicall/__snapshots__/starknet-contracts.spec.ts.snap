// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Starknet multicall > should return responses from different contract calls 1`] = `
[
  [
    "Staking Delegation Pool",
  ],
  [
    "0x0",
    "0x1c6a9f1f3f97a0a3b176fac97b69c3cd21ba31a522fca111b0e49e46975a048",
    "0x29fa8220372487cb4e1",
    "0x41a4db19d798c5601b",
    "0x0",
    "0x0",
    "0x1",
  ],
  [
    "0x0",
    "0xd3b910d8c528bf0216866053c3821ac6c97983dc096bff642e9a3549210ee7",
    "0x6012516a3ae0e8e8367abdb1db76ba56f7cb221aa3c1e4c02f52ab9d4d1ebc6",
    "0x1",
    "0x175a7be10a64c27440000",
    "0x24bbe83f0dac46c56c50",
    "0x0",
    "0x2cb02c72e8a0975e69e88298443e984d965a49eab38f5bdde1f5072daa09cfe",
    "0x4054bc140aecf17b7b7c48",
    "0x0",
  ],
]
`;
